src/cache_cf.cc

```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
-        file = strwordtok(nullptr, &saveptr);
+    while (auto path = NextWordWhileRemovingDoubleQuotesAndBackslashesInsideThem(tk)) {
+        error_count += parseOneConfigFile(path->c_str(), depth);
     }
 #endif /* HAVE_GLOB */
     return error_count;
 }
 
+/// Replaces all occurrences of macroName in buf with macroValue. When looking
+/// for the next macroName occurrence, this one-scan algorithm does not revisit
+/// previously scanned buf areas and does not visit replaced values.
 static void
-ReplaceSubstr(char*& str, int& len, unsigned substrIdx, unsigned substrLen, const char* newSubstr)
+SubstituteMacro(SBuf &buf, const SBuf &macroName, const SBuf &macroValue)
 {
-    assert(str != nullptr);
-    assert(newSubstr != nullptr);
-
-    unsigned newSubstrLen = strlen(newSubstr);
-    if (newSubstrLen > substrLen)
-        str = (char*)realloc(str, len - substrLen + newSubstrLen + 1);
-
-    // move tail part including zero
-    memmove(str + substrIdx + newSubstrLen, str + substrIdx + substrLen, len - substrIdx - substrLen + 1);
-    // copy new substring in place
-    memcpy(str + substrIdx, newSubstr, newSubstrLen);
+    SBuf remainingInput = buf;
+    buf.clear();
+    while (!remainingInput.isEmpty()) {
+        const auto pos = remainingInput.find(macroName);
+        if (pos == SBuf::npos) {
+            buf.append(remainingInput);
+            return;
+        }
 
-    len = strlen(str);
+        buf.append(remainingInput.substr(0, pos));
+        buf.append(macroValue);
+        remainingInput.chop(pos + macroName.length());
+    }
 }
 
 static void
-SubstituteMacro(char*& line, int& len, const char* macroName, const char* substStr)
+ProcessMacros(SBuf &buf)
 {
-    assert(line != nullptr);
-    assert(macroName != nullptr);
-    assert(substStr != nullptr);
-    unsigned macroNameLen = strlen(macroName);
-    while (const char* macroPos = strstr(line, macroName)) // we would replace all occurrences
-        ReplaceSubstr(line, len, macroPos - line, macroNameLen, substStr);
+    static const auto macroServiceName = SBuf("${service_name}");
+    static const auto macroProcessName = SBuf("${process_name}");
+    static const auto macroProcessNumber = SBuf("${process_number}");
+    static const auto kidIdentifier = ToSBuf(KidIdentifier);
+    SubstituteMacro(buf, macroServiceName, service_name);
+    SubstituteMacro(buf, macroProcessName, TheKidName);
+    SubstituteMacro(buf, macroProcessNumber, kidIdentifier);
 }
 
-static void
-ProcessMacros(char*& line, int& len)
+/// extracts all leading space characters (if any)
+/// \returns whether at least one character was extracted
+static bool
+SkipOptionalSpace(Parser::Tokenizer &tk)
 {
-    SubstituteMacro(line, len, "${service_name}", service_name.c_str());
-    SubstituteMacro(line, len, "${process_name}", TheKidName.c_str());
-    SubstituteMacro(line, len, "${process_number}", xitoa(KidIdentifier));
+    return tk.skipAll(CharacterSet::WSP);
 }
```

Comment:
Calling a one-line function is just as much duplicate code as writing its contents in full.

> This function should not be inlined because inlining it will duplicate its code 

"duplicate code" arguments are only relevant to lines which are complex (eg math equations).

One-line functions/methods that simply call another function/method such as this one are called "trampolines" and best practice is to *avoid* use of trampolines whenever possible.
> and obscure its high-level intent/logic in many callers.

The "high level intent" is clear from the syntax and parameters of `skipAll(CharacterSet::WSP);`.
This being a separate function is redundant (as is the new documentation for it).

> 
> In most cases, we should not inline, regardless of the number of lines a function has. There are exceptions to that rule of thumb, of course, but this particular function is not exceptional.

Reply:
> Calling a one-line function is just as much duplicate code as writing its contents in full.

That assertion is false for at least two independent reasons:

* That encapsulated code line contains more details than a function name. It contains the choice of delimiters to use when skipping space.
* Using the same context-specific wrapper function (name) in multiple places carries more information than calling a low level skipAll() method would. Knowing that all these places are related and all of them skip optional space (rather than, say, required space) helps read and maintain code.


> > This function should not be inlined because inlining it will duplicate its code
> 
> "duplicate code" arguments are only relevant to lines which are complex (eg math equations).

That assertion is false because complexity is not the only thing we do not want to duplicate. Duplicating simple decisions is also a negative.

> The "high level intent" is clear from the syntax and parameters of `skipAll(CharacterSet::WSP);`.

Does that call skip optional or required space? What other places should be updated if I want to update WSP in one particular place? Encapsulating that call helps answer these and other questions. Yes, those answers can usually be derived without encapsulation, but at a higher cost (sometimes significantly higher cost!).

> > In most cases, we should not inline, regardless of the number of lines a function has. There are exceptions to that rule of thumb, of course, but this particular function is not exceptional.
> 
> I Disagree, one-line functions are all special-case for that rule of thumb. 

I see no good reason to make one-line functions a special case. The onus is on you to prove such an exception is needed (just stating that it is needed is not enough).

N.B. Absence of an exception to "do not inline unless you have a very good reason" rule of thumb does not mean that one-line functions are _always_ better than duplicated inlined code, of course.

> This one is also an exceptional in addition to that one-line nature:
> 
> FYI: One-line functions/methods that simply call another function/method (such as this one does) are called "trampolines" and best practice is to _avoid_ use of trampolines whenever possible.

The term "[trampoline function](https://en.wikipedia.org/wiki/Trampoline_(computing))" means quite different things in various programming languages/contexts, but it is _not_ used to describe "a function calling another function" in any relevant context I can find. The term is usually related to functions that define/contain (not just call!) other functions or receive other functions as parameters/state. None of that applies here, of course.